source/darwin/nw_socket.c

```
@@ -1255,27 +1658,117 @@ static void s_handle_write_fn(
+            }
+        }
+    }
+
+    return AWS_OP_SUCCESS;
+}
+
 static int s_socket_connect_fn(
     struct aws_socket *socket,
-    const struct aws_socket_endpoint *remote_endpoint,
-    struct aws_event_loop *event_loop,
-    aws_socket_on_connection_result_fn *on_connection_result,
+    struct aws_socket_connect_options *socket_connect_options,
     void *user_data) {
     struct nw_socket *nw_socket = socket->impl;
 
-    AWS_FATAL_ASSERT(event_loop);
-    AWS_FATAL_ASSERT(!socket->event_loop);
+    const struct aws_socket_endpoint *remote_endpoint = socket_connect_options->remote_endpoint;
+    struct aws_event_loop *event_loop = socket_connect_options->event_loop;
+    aws_socket_on_connection_result_fn *on_connection_result = socket_connect_options->on_connection_result;
+    aws_socket_retrieve_tls_options_fn *retrieve_tls_options = socket_connect_options->retrieve_tls_options;
+
+    AWS_ASSERT(event_loop);
+    AWS_FATAL_ASSERT(on_connection_result);
 
     AWS_LOGF_DEBUG(
         AWS_LS_IO_SOCKET, "id=%p handle=%p: beginning connect.", (void *)socket, socket->io_handle.data.handle);
 
-    // Apple Network Framework uses a connection based abstraction on top of the UDP layer. We should always do an
-    // "connect" action after aws_socket_init() regardless it's a UDP socket or a TCP socket.
-    AWS_FATAL_ASSERT(on_connection_result);
+    if (socket->event_loop) {
+        return aws_raise_error(AWS_IO_EVENT_LOOP_ALREADY_ASSIGNED);
+    }
+
+    if (retrieve_tls_options != NULL) {
+        struct aws_tls_connection_context tls_connection_context;
+        AWS_ZERO_STRUCT(tls_connection_context);
+        retrieve_tls_options(&tls_connection_context, user_data);
+
+        if (s_setup_tls_options_from_context(nw_socket, &tls_connection_context)) {
+            AWS_LOGF_ERROR(
+                AWS_LS_IO_SOCKET, "id=%p: Error encounterd during setup of tls options from context.", (void *)socket);
+            return aws_last_error();
```

We don't want to return the error code here, we want to return AWS_OP_ERR.